### PR TITLE
feat: Service API keys + meeting creation for Sentinel

### PR DESCRIPTION
**Pull request by christosporios**
## Summary
- Adds service API key infrastructure (create, validate, revoke with SHA-256 hashing) so Sentinel can authenticate via `Authorization: Bearer <key>`
- Meeting creation now supports optional `meetingId` (auto-generated with `_2` collision handling), `processAgenda: true` to auto-trigger agenda parsing, and date range filtering on GET
- Admin UI at `/admin/settings/api-keys` for key lifecycle management
- "Run pipeline" checkbox (default on) in the meeting creation form

## Test plan
- [x] Create API key via admin UI, copy raw key
- [x] Use key to list cities, list meetings, create meeting with auto-generated ID
- [x] Verify `meetingId` collision handling (`apr20_2026` → `apr20_2026_2`)
- [x] Verify `includeUnreleased=true` is auth-gated (401 without token)
- [x] Revoke key, confirm subsequent requests return 401
- [x] Test `processAgenda: true` with a real agenda PDF URL
- [x] Verify "Run pipeline" checkbox in the UI form

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Introduces a new service authentication path (Bearer API keys) that effectively grants superadmin-equivalent access to multiple endpoints, plus new DB storage for keys. Changes touch auth, meeting creation, and upload flows, so mistakes could widen access or leak secrets.
> 
> **Overview**
> Enables programmatic access for trusted services via **service API keys** (Bearer tokens), including a new `ServiceApiKey` table, key hashing/storage, last-used tracking, and superadmin-only endpoints/UI to create/list/revoke keys (raw key shown only once).
> 
> Updates city meetings and upload APIs to accept **service-or-session auth** via `withServiceOrUserAuth`, adds optional `meetingId` auto-generation with collision handling, supports `processAgenda` to auto-trigger agenda parsing on create, and expands meetings listing with `from`/`to` filtering plus auth-gated `includeUnreleased`.
> 
> Refactors agenda processing by extracting core logic into `requestProcessAgendaInternal` (no auth) and adds a meeting-creation form checkbox + i18n strings to opt into running the pipeline after creation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b7f9030d4c93c5d49bd213c455036a86c001c00a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

This PR introduces service API key infrastructure (SHA-256 hashed Bearer tokens) for programmatic access by Sentinel, with a superadmin UI and lifecycle management (create/list/revoke). It also extends the meetings API with optional `meetingId` auto-generation, `processAgenda` triggering, date-range filtering, and auth-gated `includeUnreleased`. Previous review rounds addressed the TOCTOU race condition, the stale `meetingId` after retry, redundant DB round-trips, and the `hashedKey` exposure — all confirmed fixed in this revision.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all previously flagged critical issues are resolved and only a minor P2 remains.

All P0/P1 findings from prior review rounds (TOCTOU race, stale `meetingId` after retry, `hashedKey` exposure, redundant DB round-trip) have been fixed. The one remaining comment is P2: a Prisma P2025 on the DELETE handler returning 500 instead of 404 in the edge case of revoking an already-deleted key, which is a superadmin-only path and low impact.

src/app/api/admin/api-keys/[id]/route.ts — minor P2025 handling gap

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/lib/auth.ts | Adds `withServiceOrUserAuth` — Bearer token checked first (invalid token rejects without session fallback), then falls back to `checkUserAuthorization`. Second `getCurrentUser()` call acknowledged with TODO. Logic is sound. |
| src/lib/db/apiKeys.ts | SHA-256 hashing, `base64url` encoding for raw keys, fire-and-forget `lastUsedAt` update, `select`-based listing (no hash leak). Solid implementation. |
| src/app/api/admin/api-keys/[id]/route.ts | Superadmin DELETE handler; will 500 instead of 404 on missing key ID because Prisma P2025 is not mapped to a 404 in `handleApiError`. |
| src/app/api/cities/[cityId]/meetings/route.ts | POST now accepts optional `meetingId`, auto-generates via `generateUniqueMeetingId`, retries on P2002, updates `meetingId` before side-effects. GET adds `from`/`to` filtering and auth-gated `includeUnreleased`. All previously flagged race-condition issues are addressed. |
| src/lib/db/meetings.ts | Adds `createCouncilMeetingDirect` (auth-free) and `generateUniqueMeetingId` with startsWith prefix query + Set-based exact matching. Collision loop capped at 20 suffixes. Date filtering added to `getCouncilMeetingsForCity`. |
| src/lib/tasks/processAgendaInternal.ts | Correctly extracted core agenda-processing logic into a non-"use server" module. Auth-free by design; callers are responsible for prior authorization. |
| src/components/meetings/AddMeetingForm.tsx | Adds "Run pipeline" checkbox (default on, disabled without agenda URL). Visual state correctly derived from `field.value && hasAgenda`; form state and backend behaviour are consistent. |
| src/lib/utils/meetingId.ts | Extracts `formatDateAsMeetingId` with Athens timezone pinning. Well-tested for edge cases (midnight UTC, single-digit day). |
| prisma/migrations/20260416000002_add_service_api_keys/migration.sql | Creates `ServiceApiKey` table with unique index on `hashedKey`, FK to `User` with RESTRICT on delete. Schema is clean and minimal. |
| src/app/api/upload/presigned-url/route.ts | Replaces `isUserAuthorizedToEdit` with `withServiceOrUserAuth`, properly catching `ApiError` to return the correct HTTP status. Semantic change from 403 to 401/403 depending on auth failure type is acceptable. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant S as Sentinel / Client
    participant MW as withServiceOrUserAuth
    participant AK as validateServiceApiKey
    participant DB as Prisma (DB)
    participant PA as requestProcessAgendaInternal

    S->>MW: POST /api/cities/{cityId}/meetings Authorization: Bearer sk_...
    MW->>AK: validateServiceApiKey(token)
    AK->>DB: findUnique(hashedKey)
    DB-->>AK: ServiceApiKey row
    AK-->>MW: { name: Sentinel }
    MW-->>S: { type: service, keyName }

    Note over S,DB: Auth passed — proceed with meeting creation
    S->>DB: generateUniqueMeetingId(cityId, date)
    DB-->>S: apr20_2026
    S->>DB: createCouncilMeetingDirect(...)
    alt P2002 collision
        DB-->>S: PrismaClientKnownRequestError
        S->>DB: generateUniqueMeetingId (retry)
        DB-->>S: apr20_2026_2
        S->>DB: createCouncilMeetingDirect(...)
    end
    DB-->>S: CouncilMeeting

    opt processAgenda=true and agendaUrl present
        S->>PA: requestProcessAgendaInternal(agendaUrl, meetingId, cityId)
        PA->>DB: findUnique(councilMeeting)
        PA-->>S: Task status
    end

    S-->>S: 201 meeting + processAgendaStatus
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/components/meetings/AddMeetingForm.tsx`, line 76-79 ([link](https://github.com/schemalabz/opencouncil/blob/f5f1de5d30c8dba89d51332aeed340526df74da6/src/components/meetings/AddMeetingForm.tsx#L76-L79)) 

   <a href="#">P2</a> **Duplicate `formatDateAsMeetingId` not consolidated**

   This PR extracts `formatDateAsMeetingId` into `src/lib/utils/meetingId.ts` specifically so it can be shared, but the local copy inside this component was not replaced. The two implementations are currently identical, but they can drift. The component can safely import the shared version (it's a pure function with no server-only dependencies).

   Remove the inline definition and add the import at the top of the file:
   ```
   import { formatDateAsMeetingId } from "@/lib/utils/meetingId"
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/components/meetings/AddMeetingForm.tsx
   Line: 76-79

   Comment:
   **Duplicate `formatDateAsMeetingId` not consolidated**

   This PR extracts `formatDateAsMeetingId` into `src/lib/utils/meetingId.ts` specifically so it can be shared, but the local copy inside this component was not replaced. The two implementations are currently identical, but they can drift. The component can safely import the shared version (it's a pure function with no server-only dependencies).

   Remove the inline definition and add the import at the top of the file:
   ```
   import { formatDateAsMeetingId } from "@/lib/utils/meetingId"
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/app/api/admin/api-keys/[id]/route.ts
Line: 15-21

Comment:
**Revoking a non-existent key returns 500 instead of 404**

`prisma.serviceApiKey.update` throws a `PrismaClientKnownRequestError` with code `P2025` when the record doesn't exist. `handleApiError` doesn't recognise Prisma errors and falls back to returning the raw `error.message` with a 500 status, rather than a clean 404. In practice this only affects stale UI (e.g. two admins revoking the same key concurrently), but it's noisy and leaks the internal Prisma error message.

```suggestion
    try {
        await revokeServiceApiKey(params.id);
        return NextResponse.json({ success: true });
    } catch (error) {
        if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2025') {
            return NextResponse.json({ error: "API key not found" }, { status: 404 });
        }
        return handleApiError(error, "Failed to revoke API key");
    }
```

(Requires adding `import { Prisma } from '@prisma/client';` at the top.)

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/lib/db/meetings.ts
Line: 1143-1155

Comment:
**`startsWith` filter may include false-positive rows from other meetings**

The query `id: { startsWith: baseId }` is intended to find all collision variants for a given date slug (e.g. `apr20_2026`, `apr20_2026_2`, …). However, if another meeting in the same city has an ID that happens to share the same prefix for unrelated reasons (e.g. a manually created `apr20_2026_custom`), it will be pulled into `existingIds`. Because the loop then only checks for `apr20_2026_2`, `apr20_2026_3`, etc. (exact match), the extra row is benign — it just adds a tiny overhead.

More concretely: the loop will still skip over the manually-created ID and return the first `_N` suffix that is absent, so there is no correctness issue. Worth noting in a comment so future readers don't second-guess the `Set` lookup strategy.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (8): Last reviewed commit: ["fix: move generateUniqueMeetingId to db ..."](https://github.com/schemalabz/opencouncil/commit/f47441ea158cdd5f4541f7232ef6c28846b14b53) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28209558)</sub>

### CODE REVIEW

File: src/lib/db/apiKeys.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+export async function getServiceApiKeys() {
+    return prisma.serviceApiKey.findMany({
+        orderBy: { createdAt: 'desc' },
+        include: {
+            createdBy: {
+                select: { name: true, email: true },
+            },
+        },
+    });
+}
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/db/apiKeys.ts:60]: Fixed — switched from `include` to `select` so `hashedKey` is never returned in the API response.

File: src/lib/auth.ts [hunk 2]
```diff
@@ -147,6 +150,43 @@ export async function isUserAuthorizedToEdit({
...
+    }
+
+    // Fall back to session auth
+    await withUserAuthorizedToEdit({ cityId });
+
+    const user = await getCurrentUser();
+    if (!user) {
+        throw new UnauthorizedError("Session expired");
+    }
+    return { type: 'user', userId: user.id };
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/auth.ts]: Addressed — now calls `checkUserAuthorization` directly instead of going through `withUserAuthorizedToEdit`. There's still one extra `getCurrentUser()` call to get the user ID for the return value; left a TODO to refactor `checkUserAuthorization` to return the user.

File: src/lib/utils/meetingId.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    for (let suffix = 2; suffix <= 20; suffix++) {
+        const candidateId = `${baseId}_${suffix}`;
+        if (!existingIds.has(candidateId)) {
+            return candidateId;
+        }
+    }
+
+    throw new Error(`Could not generate unique meeting ID for ${cityId} on ${baseId} — too many meetings on this date`);
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/utils/meetingId.ts:47]: Fixed — route now catches P2002 unique constraint violations and retries with a freshly generated ID.

File: src/lib/auth.ts [hunk 4]
```diff
@@ -147,6 +150,43 @@ export async function isUserAuthorizedToEdit({
...
+        const apiKey = await validateServiceApiKey(token);
+        if (apiKey) {
+            return { type: 'service', keyName: apiKey.name };
+        }
+        // Invalid bearer token — don't fall through to session auth
+        throw new UnauthorizedError("Invalid API key");
+    }
+
+    // Fall back to session auth
+    await withUserAuthorizedToEdit({ cityId });
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/auth.ts]: Fixed — `withServiceOrUserAuth` now throws `UnauthorizedError` (from `ApiError` hierarchy) for both bearer and session auth failures, so `handleApiError` returns proper 401s.

File: src/lib/tasks/processAgenda.ts [hunk 5]
```diff
@@ -9,12 +9,16 @@ import { getAllTopics } from "../db/topics";
 import { getPartyFromRoles, getRoleNameForPerson } from "../utils";
 import { getPeopleForMeeting } from "../db/people";
 
-export async function requestProcessAgenda(agendaUrl: string, councilMeetingId: string, cityId: string, {
+/**
+ * Core processAgenda logic without auth checks.
+ * Called by the route handler (which already authenticated the request)
+ * and by the user-facing wrapper below.
+ */
+export async function requestProcessAgendaInternal(agendaUrl: string, councilMeetingId: string, cityId: string, {
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/tasks/processAgenda.ts]: Fixed — moved `requestProcessAgendaInternal` to a separate `processAgendaInternal.ts` file without the `"use server"` directive. No longer callable as a Server Action.

File: src/lib/utils/meetingId.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/utils/meetingId.ts:48]: Fixed — P2002 retry logic added in the route handler.

File: src/app/api/cities/[cityId]/meetings/route.ts [hunk 7]
```diff
@@ -37,32 +40,67 @@ const getMeetingsQuerySchema = z.object({
...
+                : createCouncilMeeting(buildMeetingData(id));
+
+        // Retry with a fresh ID on unique constraint violation (TOCTOU race)
+        let meeting;
+        try {
+            meeting = await createMeeting(meetingId);
+        } catch (error) {
+            const isUniqueViolation = error instanceof Error && 'code' in error && (error as { code: string }).code === 'P2002';
+            if (!isUniqueViolation || providedMeetingId) throw error;
+            const retryId = await generateUniqueMeetingId(cityId, date);
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/api/cities/[cityId]/meetings/route.ts]: Fixed — `meetingId` is now `let` and updated on retry, so all downstream side-effects (Discord, Calendar, processAgenda) use the correct ID.

File: src/app/api/cities/[cityId]/meetings/route.ts [hunk 8]
```diff
@@ -37,32 +40,67 @@ const getMeetingsQuerySchema = z.object({
...
-        await withUserAuthorizedToEdit({ cityId: params.cityId });
+        const authResult = await withServiceOrUserAuth(request, { cityId: params.cityId });
         const body = await request.json();
-        const { name, name_en, date, youtubeUrl, agendaUrl, meetingId, administrativeBodyId } = meetingSchema.parse(body);
+        const { name, name_en, date, youtubeUrl, agendaUrl, meetingId: providedMeetingId, administrativeBodyId, processAgenda } = meetingSchema.parse(body);
         const cityId = params.cityId;
 
-        const meeting = await createCouncilMeeting({
+        // Auto-generate meetingId if not provided
+        const meetingId = providedMeetingId || await generateUniqueMeetingId(cityId, date);
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/api/cities/[cityId]/meetings/route.ts]: Fixed — same as Greptile's comment above. `meetingId` updated on retry path.

File: .claude/scheduled_tasks.lock [hunk 9]
```diff
@@ -0,0 +1 @@
+{"sessionId":"b02fc5ad-6b6f-414f-b263-45cc03a83c9c","pid":61125,"acquiredAt":1776088102969}
```
[the remark this answers is not in the document]
Comment (christosporios) [reply .claude/scheduled_tasks.lock]: Fixed — removed from git and added to `.gitignore`.

File: src/app/api/upload/presigned-url/route.ts [hunk 10]
```diff
@@ -92,15 +92,11 @@ export async function POST(request: NextRequest) {
...
-        // Check user authorization
-        // If cityId is provided in config, check city-specific permissions
-        // Otherwise, check general upload permissions
+        // Check authorization via service API key or user session
         const cityId = config?.cityId
-        const authorizedToEdit = await isUserAuthorizedToEdit(
-            cityId ? { cityId } : {}
-        )
-
-        if (!authorizedToEdit) {
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/api/upload/presigned-url/route.ts:103]: Fixed — narrowed catch to only handle `ApiError` instances, re-throws unexpected errors.

File: src/app/api/cities/[cityId]/meetings/route.ts [hunk 11]
```diff
@@ -37,32 +40,67 @@ const getMeetingsQuerySchema = z.object({
...
+
+        // Retry with a fresh ID on unique constraint violation (TOCTOU race)
+        let meeting;
+        try {
+            meeting = await createMeeting(meetingId);
+        } catch (error) {
+            const isUniqueViolation = error instanceof Error && 'code' in error && (error as { code: string }).code === 'P2002';
+            if (!isUniqueViolation || providedMeetingId) throw error;
+            meetingId = await generateUniqueMeetingId(cityId, date);
+            meeting = await createMeeting(meetingId);
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/app/api/cities/[cityId]/meetings/route.ts]: Fixed — retry path now uses direct prisma create (bypassing `createCouncilMeeting`'s error wrapping that loses the P2002 code). Auth was already verified at route level.

File: src/lib/utils/meetingId.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ * This logic was originally in AddMeetingForm (client-side).
+ * Extracted here so it can be used server-side for auto-generation.
+ */
+export function formatDateAsMeetingId(date: Date): string {
+    return date
+        .toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' })
+        .toLowerCase()
+        .replace(/\s/g, '')
+        .replace(',', '_');
+}
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/utils/meetingId.ts:16]: Fixed — pinned `timeZone: 'Europe/Athens'` in the server-side `formatDateAsMeetingId`.

[the remark this answers is not in the document]
Comment (christosporios) [reply src/lib/utils/meetingId.ts:16]: Intentional — the client component (`AddMeetingForm.tsx`) can't import from `meetingId.ts` because it pulls in `prisma` at the top level, which would break the client bundle. The duplication is 3 lines and both copies are trivial. Not worth restructuring imports for.

**Comment by kouloumos** [comment]
Rebased onto latest main to pick up the deployment fix (0ac66668) that we pushed to address the issue we hit on this PR. Force-pushed the result — the original branch state is preserved at [`backup/pr-315-original`](https://github.com/kouloumos/opencouncil/tree/backup/pr-315-original) just in case.

File: src/app/api/cities/[cityId]/meetings/route.ts [hunk 13]
```diff
@@ -37,32 +40,69 @@ const getMeetingsQuerySchema = z.object({
...
+        // User auth: delegate to createCouncilMeeting which re-checks session auth.
+        let meeting;
+        try {
+            meeting = authResult.type === 'service'
+                ? await createMeetingDirect(meetingId)
+                : await createCouncilMeeting(buildMeetingData(meetingId));
+        } catch (error) {
+            // Retry with a fresh ID on unique constraint violation (TOCTOU race).
+            // Always use direct create for retry — auth was already verified above.
+            const isUniqueViolation = error instanceof Error && 'code' in error && (error as { code: string }).code === 'P2002';
```
Comment (kouloumos) [inline src/app/api/cities/[cityId]/meetings/route.ts]: The retry never fires for user-auth requests — `createCouncilMeeting` wraps all Prisma errors in a generic `new Error('Failed to create council meeting')`, stripping the `P2002` code. Using a single creation path (with the auth-skip option mentioned above) would fix this — the retry would always see the raw Prisma error.

File: prisma/schema.prisma [hunk 14]
```diff
@@ -763,6 +764,21 @@ model Administers {
...
+  hashedKey   String    @unique // SHA-256 hash of the API key
+  keyPrefix   String    // First 8 chars for display (e.g. "sk_a1b2c3")
+  createdAt   DateTime  @default(now())
+  lastUsedAt  DateTime?
+  revokedAt   DateTime?
+
+  createdBy   User      @relation(fields: [createdById], references: [id])
+  createdById String
+
+  @@index([hashedKey])
```
Comment (kouloumos) [inline prisma/schema.prisma]: `@unique` already creates a B-tree index. This `@@index` creates a redundant non-unique index on the same column — wasted storage and write overhead with no read benefit.

Also, the `keyPrefix` comment says "First 8 chars" but the code uses `substring(0, 10)` — the comment and the example don't match the actual length stored.

File: src/app/api/cities/[cityId]/meetings/route.ts (hunk 13 above)
Comment (kouloumos) [inline src/app/api/cities/[cityId]/meetings/route.ts]: Instead of `error as { code: string }` use `error instanceof Prisma.PrismaClientKnownRequestError && error.code === 'P2002'` for proper type narrowing.

File: src/app/api/admin/api-keys/route.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+export async function POST(request: Request) {
+    const user = await getCurrentUser();
+    if (!user?.isSuperAdmin) {
+        return NextResponse.json({ error: "Unauthorized" }, { status: 401 });
+    }
+
+    try {
+        const { name } = await request.json();
+        if (!name || typeof name !== 'string' || name.trim().length === 0) {
```
Comment (kouloumos) [inline src/app/api/admin/api-keys/route.ts]: Manual validation instead of Zod, inconsistent with other routes. No max-length check. Use `z.object({ name: z.string().min(1).max(100).trim() })`.

File: src/app/api/admin/api-keys/[id]/route.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+import { revokeServiceApiKey } from "@/lib/db/apiKeys";
+import { handleApiError } from "@/lib/api/errors";
+
+export async function DELETE(
+    _request: Request,
+    { params }: { params: { id: string } }
+) {
+    const user = await getCurrentUser();
+    if (!user?.isSuperAdmin) {
+        return NextResponse.json({ error: "Unauthorized" }, { status: 401 });
```
Comment (kouloumos) [inline src/app/api/admin/api-keys/[id]/route.ts]: A non-superadmin user IS authenticated, just not authorized — this should be 403. Same issue in the GET/POST route. Existing admin routes like `users/[userId]/route.ts` distinguish these correctly.

File: src/app/api/cities/[cityId]/meetings/route.ts [hunk 17]
```diff
@@ -37,32 +40,69 @@ const getMeetingsQuerySchema = z.object({
...
+                include: { administrativeBody: true },
+            });
+
+        // Service auth: route already verified the API key, so create directly.
+        // User auth: delegate to createCouncilMeeting which re-checks session auth.
+        let meeting;
+        try {
+            meeting = authResult.type === 'service'
+                ? await createMeetingDirect(meetingId)
+                : await createCouncilMeeting(buildMeetingData(meetingId));
```
Comment (kouloumos) [inline src/app/api/cities/[cityId]/meetings/route.ts]: The route creates two paths for meeting creation: `createMeetingDirect` (raw Prisma call) for service auth, and `createCouncilMeeting` (from `src/lib/db/meetings.ts`) for user auth. This is needed because `createCouncilMeeting` calls `withUserAuthorizedToEdit` internally, which checks the session — and service keys don't have sessions.

The problem is that having two paths means they can silently diverge. All database mutations in this codebase go through the centralized functions in `src/lib/db/` — `createCouncilMeeting`, `editCouncilMeeting`, `deleteCouncilMeeting`, etc. If `createCouncilMeeting` gains logic later (audit logging, validation, notifications), the service path skips it.

It also causes the TOCTOU retry bug: `createCouncilMeeting` wraps all Prisma errors in a generic `new Error('Failed to create council meeting')`, stripping the `P2002` error code. The retry checks for `P2002`, so it never fires on the user-auth path.

Both issues go away if `createCouncilMeeting` accepts a `{ skipAuth: true }` option and re-throws the original error instead of wrapping it. The route already verified auth via `withServiceOrUserAuth` above, so the internal auth check is redundant for both paths. Then:
- Both auth types go through the same function
- The retry sees the raw Prisma error and works correctly
- `createMeetingDirect` can be dropped

The re-throw is safe — no caller in the codebase catches on the `'Failed to create council meeting'` message.

**Review by kouloumos — COMMENTED** [review]
The key management infrastructure looks good -- SHA-256 hashing, show-once raw key, Bearer flow with no fall-through on invalid token. The processAgenda extraction into a non-`"use server"` module is the right call. A few things to address, left inline.

The main thing to look at is the meeting creation path -- there's a dual-path issue where service auth bypasses `createCouncilMeeting` and uses a direct Prisma call, which also causes the TOCTOU retry to be dead code on the user-auth path. Left a detailed comment inline with a suggested fix. The rest are smaller items, also inline.

**Comment by kouloumos** [comment]
I pushed four commits on top with fixes — please verify before we merge:

**f2b83e0a — Unify meeting creation path.** The route had two creation paths: `createMeetingDirect` (raw Prisma) for service auth, `createCouncilMeeting` for user auth. This meant the service path bypassed `src/lib/db/meetings.ts`, so any logic added to `createCouncilMeeting` later would silently not apply to service-created meetings. It also caused the TOCTOU retry to be dead code on the user-auth path — `createCouncilMeeting` wraps all Prisma errors in a generic `Error`, stripping the `P2002` code the retry needs.

Extracted `createCouncilMeetingDirect` in the db layer (no auth, raw Prisma errors) and have `createCouncilMeeting` delegate to it after auth. The route now uses `createCouncilMeetingDirect` for both auth types since `withServiceOrUserAuth` already verified auth above. Also switched to `Prisma.PrismaClientKnownRequestError` for proper P2002 type narrowing.

**e7445b97 — Move `generateUniqueMeetingId` to db layer, deduplicate `formatDateAsMeetingId`.** `generateUniqueMeetingId` queries the database so it belongs in `src/lib/db/meetings.ts`, not `src/lib/utils/`. Moving it out makes `meetingId.ts` a pure utility with no Prisma import, which lets `AddMeetingForm` import `formatDateAsMeetingId` directly instead of maintaining a local copy. The local copy was missing `timeZone: 'Europe/Athens'`, so the client preview could diverge from the server-generated ID for users in other timezones.

**1037a224 — Tests.** `formatDateAsMeetingId` (basic formatting, single-digit days, timezone edge cases) and `generateUniqueMeetingId` (no collision, suffixes, gaps, exhaustion at `_20`).

**605db25b — Smaller fixes:**
- Removed redundant `@@index([hashedKey])` — `@unique` already creates an index. Fixed the `keyPrefix` comment to match actual length.
- Switched admin API key routes to Zod validation (consistent with other routes, adds max-length).
- Return 403 instead of 401 for authenticated non-superadmin users, matching existing admin route conventions.

**Comment by kouloumos** [comment]
Rebased your commits on top of the current main and force-pushed the result.

One change to note: the migration was renamed from `20260413130227_add_service_api_keys` to `20260416000002_add_service_api_keys` so it applies after the migrations that landed on main in the meantime.

The original branch state before the rebase is preserved at [`backup/pr-315-original`](https://github.com/kouloumos/opencouncil/tree/backup/pr-315-original), so nothing is lost.

**Comment by kouloumos** [comment]
Tested the Sentinel API endpoints e2e on the preview (`pr-315.preview.opencouncil.gr`).

**API key creation**
- Created key via admin UI — worked, key displayed once
- Created key directly in DB for initial testing — also validated correctly
- `lastUsedAt` updated on both keys after use

**Bearer auth**
- Valid key (both UI-created and DB-inserted) → 200/201 on all protected endpoints
- Invalid key → 401 `Invalid API key`
- No auth on protected route → 401 `Not authorized`

**`POST /api/cities/{cityId}/meetings`**
- Created meeting with auto-generated `meetingId` (`apr21_2026`) — correct format
- Created a second meeting for the same date — collision handling produced `apr21_2026_2`
- Meeting with `processAgenda: true` + `agendaUrl`: full pipeline executed — built request body with people, topics, callback URL, and POSTed to task backend. Failed at the task server with 403 (preview `TASK_API_KEY` mismatch), but the code path is correct and the response gracefully returns `processAgendaStatus: "failed"`

**`GET /api/cities/{cityId}/meetings`**
- `includeUnreleased=true` with Bearer auth returned unreleased meetings
- `from`/`to` date filtering worked

**`POST /api/upload/presigned-url`**
- Auth passed (Bearer accepted), but returned 500 due to invalid `DO_SPACES_ENDPOINT` on preview — not a code issue
